README.md
# ui-check

[vscode插件]这是一个HTML边界检测生成工具

## Features

很多情况下，HTML原型（小程序）都是最完美的状态，对于文字过多或者为空，图片尺寸大小等可能并没有做相关的设定，可能导致在实际上线后出现各种边界问题（比如文本溢出、高度塌陷、图片过大等等）

该插会件生成-test目录，里面包含三个文件夹，分别是`empty`、`overflow`、`random`，如下

```
./test                 -> test生成目录
  |--empty             -> 文本内容为空、图片为空的情况
  |--overflow          -> 文本内容很多、图片资源随机的情况
  |--random            -> 文本图片资源随机的情况
```

> 提示：目前支持HTML原型(.html)、小程序(.wxml)、vue项目(.vue)

## Usage

安装完以后左下角会出现`ui-check`的按钮

![ui-check](https://imgservices-1252317822.image.myqcloud.com/image/20200701/dur6hskjs4.jpg)

点击该按钮，编辑器顶部会出现输入框，输入所需要检测目录，默认为`./dist/`

![ui-input](https://imgservices-1252317822.image.myqcloud.com/image/20200701/pl1875wydr.jpg)

> 提示：相对路径，比如`./`表示当前目录，`../`表示上一层目录等等

回车`Enter`，会自动生成与检测目录并列的目录，文件夹名称为`/检测目录-test`，比如检测文件夹为`/src`，生成的则是`/src-test`，同时右下角提示“目录生成完成”

![tips](https://imgservices-1252317822.image.myqcloud.com/image/20200702/6wu3tfys78.jpg)

点击打开，则会跳转到文件管理器

![file](https://imgservices-1252317822.image.myqcloud.com/image/20200702/12dymh2mki.jpg)

如果输入目录不存在，则会提示错误

![error](https://imgservices-1252317822.image.myqcloud.com/image/20200701/9t49myo3t2.jpg)

## ScreenShot

|**原始页面**|**文本内容为空、图片为空的情况**|**文本内容很多、图片资源随机的情况**|
|----|----|----|
|![原始页面](https://imgservices-1252317822.image.myqcloud.com/image/20200701/cms5zwh816.jpg)|![文本内容为空、图片为空的情况](https://imgservices-1252317822.image.myqcloud.com/image/20200701/k4k0wuhtkh.jpg)|![文本内容很多、图片资源随机的情况](https://imgservices-1252317822.image.myqcloud.com/image/20200701/km4l73m1y8.jpg)|


各种极限情况一目了然

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "ui-check" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

extension.js
const vscode = require('vscode');
const path = require('path');
const build = require('./run');

let statusBar = null;
let dist = '/dist/';
/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	console.log('Congratulations, your extension "ui-check" is now active!');

	let disposable = vscode.commands.registerCommand('ui-check.build', function () {
		vscode.window.showInputBox({ // 这个对象中所有参数都是可选参数
			password:false, // 输入内容是否是密码
			placeHolder:'输入目录', // 在输入框内的提示信息
			// ignoreFocusOut:true,
			value: dist,
			prompt:'需要检测的目录(默认为/dist/)', // 在输入框下方的提示信息
			validateInput: function(value){
				if(!value){
					return '目录不能为空'
				}
			}
		}).then((value) => {
			if(value){
				dist = value;
				build(value, (filename) => {
					const pathDist = path.join(vscode.workspace.rootPath,value,'.') + '-test';
					statusBar.text = `$(loading) 正在生成${filename}...`;
					this.timer && clearTimeout(this.timer);
					this.timer = setTimeout(() => {
						statusBar.text = '$(symbol-keyword) ui-check';
						vscode.window.showInformationMessage('目录生成完成！是否立即打开？','打开', '关闭').then(result => {
							if(result === '打开'){
								vscode.env.openExternal(vscode.Uri.parse(pathDist));
							}
						});
					},300)
				})
			}
		});
		
	});

	context.subscriptions.push(disposable);

	if(!statusBar) {
		statusBar  = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
	}
	statusBar.text = '$(symbol-keyword) ui-check';
	statusBar.show();
	statusBar.command = 'ui-check.build';
}


// this method is called when your extension is deactivated
function deactivate() {
	statusBar.dispose();
}

module.exports = {
	activate,
	deactivate
}


run.js
const vscode = require('vscode');

const fs = require('fs');

const path = require('path');
const { TextDecoder } = require('util');


/*
** 创建路径对应的文件夹（如果没有）
** @params path 目标路径
*/
const createPath = function (path) {
	// 路径有下面这几种
	// 1. /User/...      OS X
	// 2. E:/mydir/...   window
	// 3. a/b/...        下面3个相对地址，与系统无关
	// 4. ./a/b/...
	// 5. ../../a/b/...

	path = path.replace(/\\/g, '/');

	var pathHTML = '.';
	if (path.slice(0,1) == '/') {
		pathHTML = '/';
	} else if (/:/.test(path)) {
		pathHTML = '';
	}

	path.split('/').forEach(function(filename) {
		if (filename) {
			// 如果是数据盘地址，忽略
			if (/:/.test(filename) == false) {
				pathHTML = pathHTML + '/' + filename;
				// 如果文件不存在
				if(!fs.existsSync(pathHTML)) {
					console.log('路径' + pathHTML + '不存在，新建之');
					fs.mkdirSync(pathHTML);
				}
			} else {
				pathHTML = filename;
			}
		}
	});
}

/*
** 删除文件极其目录方法
** @src 删除的目录
*/
const clean = function (src) {
	if (!fs.existsSync(src)) {
		return;
	}

	// 读取目录中的所有文件/目录
	var paths = fs.readdirSync(src);

	paths.forEach(function (dir) {
		let _src = path.join(src, dir);

		let st = fs.statSync(_src);

		if (st.isFile()) {
			// 如果是文件，则删除
			fs.unlinkSync(_src);
		} else if (st.isDirectory()) {
			// 作为文件夹
			clean(_src);
		}
	});

	// 删除文件夹
	try {
		fs.rmdirSync(src);
		console.log('已清空文件夹' + src);
	} catch(e) {}
};

/*
 * 复制目录中的所有文件包括子目录
 * @param{ String } 需要复制的目录
 * @param{ String } 复制到指定的目录
 */
const copy = function (src, dst, callback, building) {
	if (!fs.existsSync(src)) {
		vscode.window.showErrorMessage(`${src}文件夹不存在！`)
		return;
	}

	// 读取目录中的所有文件/目录
	var paths = fs.readdirSync(src);
	paths.forEach(function (dir) {
		var _src = path.join(src, dir);
		var	_dst = path.join(dst, dir);

        let st = fs.statSync(_src);

		// 判断是否为文件
		if (st.isFile()) {
            if( ['.html','.vue','.wxml'].includes(path.extname(_src))){
                fs.readFile(_src,'utf8',function(err,file){
                    const filename = path.join(dst, path.basename(_src));
                    file = callback(file,dst);
                    fs.writeFile(filename,file,function(){
						building && building(filename);
                    })
                })
            }else{
                const readable = fs.createReadStream(_src);
                // 创建写入流
                const writable = fs.createWriteStream(_dst);
                // 通过管道来传输流
                readable.pipe(writable);
            }

		} else if (st.isDirectory()) {
			// 作为文件夹处理
			createPath(_dst);
			copy(_src, _dst, callback, building);
		}
	});
};


const imglist = [
    'https://imgservices-1252317822.image.myqcloud.com/image/20200701/obiovu02uy.jpg',
	'https://imgservices-1252317822.image.myqcloud.com/image/20200701/f0tutqgdiv.jpg',
	'https://imgservices-1252317822.image.myqcloud.com/image/20200701/zh6l6azztp.jpg',
	'https://imgservices-1252317822.image.myqcloud.com/image/20200701/8dzh5p5jg7.jpg',
	'https://imgservices-1252317822.image.myqcloud.com/image/20200701/scmo69275p.jpg',
	'https://imgservices-1252317822.image.myqcloud.com/image/20200701/c5g6yl6np0.jpg',
	'https://imgservices-1252317822.image.myqcloud.com/image/20200701/73dajixpc6.jpg'
]

const tasks = [
    {
        desc: '内容比较多的情况',
        dir: 'overflow/',
        times: 4,
        img: true
    },
    {
        desc: '内容为空的情况',
        dir: 'empty/',
        times: 0,
        img: false
    },
    {
        desc: '内容随机的情况',
        dir: 'random/',
        random: true,
    },
]

const randomImg = function(){
    const len = imglist.length;
    return imglist[Math.floor(Math.random()*len)];
}

const renderTxt = function(str,t){
	if(str.trim()){
		if(t.random){
			return str.repeat(Math.floor(Math.random()*4))
		}
		return str.repeat(t.times);
	}else{
		return str;
	}
}

const renderImg = function(match,source,t){
	if(t.random){
		return match.replace(source,Math.random()>.5?randomImg():null)
	}
	return match.replace(source,t.img?randomImg():null);
}

const workspace = vscode.workspace.rootPath;

const bulid = function(src,building){
	const pathSrc = path.join(workspace,src);
	const pathDist = path.join(pathSrc,'.') + '-test';
	clean(pathDist);
	tasks.forEach(function(task){
		const dist = path.join(pathDist,task.dir);
		createPath(dist);
		copy(pathSrc,dist,function(file){
			const reg_txt = /(?<=(?<!script[^>]*)>)[^<>]+(?=<(?!\/title|\/style|\/script))/g;
			const reg_img = /<(img|image) [^>]*src=['"]([^'"]+)[^>]*>/gi;
			return file.replace(reg_txt,function(str){
				return renderTxt(str,task);
			}).replace(reg_img, function (match, capture, source) {
				return renderImg(match,source,task);
			});
		},building);
	})
}

module.exports = bulid;

